### ISSUE TITLE
High CPU usage from node.mojom.NodeService

### ISSUE THREAD

**Issue by petoknm** [microsoft/vscode#213136]
Type: <b>Performance Issue</b>

This specific process always takes 100% CPU on machine:
```
/snap/code/159/usr/share/code/code --type=utility --utility-sub-type=node.mojom.NodeService --lang=en-US --service-sandbox-type=none --no-sandbox --crashpad-handler-pid=860961 --enable-crash-reporter=f5630c5d-bec1-43e9-95d6-21d878227126,no_channel --user-data-dir=/home/pmajchrak/.config/Code --standard-schemes=vscode-webview,vscode-file --secure-schemes=vscode-webview,vscode-file --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --code-cache-schemes=vscode-webview,vscode-file --shared-files=v8_context_snapshot_data:100 --field-trial-handle=0,i,4729389318826606528,17009332704461783735,262144 --enable-features=kWebSQLAccess --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --variations-seed-version
```

I looked at the different ways to profile VSCode, but they didn't show this process, so it all looked normal. Even JS profiling doesn't capture this. This process still exists even if I run VSCode with disabled extensions and the process respawns after being killed.

VS Code version: Code 1.89.1 (dc96b837cf6bb4af9cd736aa3af08cf8279f7685, 2024-05-07T05:16:23.416Z)
OS version: Linux x64 5.15.0-105-generic snap
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 9 5950X 16-Core Processor (32 x 4464)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: disabled_software<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: disabled_off|
|Load (avg)|2, 2, 2|
|Memory (System)|62.70GB (52.78GB free)|
|Process Argv|--no-sandbox --force-user-env --crash-reporter-id 18e3b18c-bfb2-4aff-9683-a7fc868c85da|
|Screen Reader|no|
|VM|0%|
|DESKTOP_SESSION|ubuntu|
|XDG_CURRENT_DESKTOP|Unity|
|XDG_SESSION_DESKTOP|ubuntu|
|XDG_SESSION_TYPE|x11|
</details><details>
<summary>Process Info</summary>

```
CPU %	Mem MB	   PID	Process
    0	   128	860943	code main
    0	     0	860945	   zygote
    0	     0	860946	   zygote
    0	   193	860993	window [1] (kas-test - Visual Studio Code)
    0	   321	876526	     window
    0	     0	876605	     utility-process
    0	    64	876682	window [3] (Issue Reporter)
    0	   257	860973	   gpu-process
    0	    64	860980	   utility-network-service
    0	    64	861272	shared-process
    0	     0	876715	     /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0	    64	861323	ptyHost
    3	  2183	874163	fileWatcher [1]
    0	    64	874175	extensionHost [1]
```

</details>
<details>
<summary>Workspace Info</summary>

```
|  Window (kas-test - Visual Studio Code)
|    Folder (kas-test): more than 24175 files
|      File types: siginfo(4707) ipk(2493) zst(2272) pyc(1408) py(1405)
|                  pyi(557) json(382) so(226) html(118) test(71)
|      Conf files: makefile(3) settings.json(2) dockerfile(2);
```

</details>
Extensions: none<details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vspor879:30202332
vspor708:30202333
vspor363:30204092
tftest:31042121
vswsl492cf:30256860
vscoreces:30445986
vscod805:30301674
binariesv615:30325510
vsaa593cf:30376535
py29gd2263:31024239
c4g48928:30535728
azure-dev_surveyone:30548225
962ge761:30959799
pythongtdpath:30769146
welcomedialog:30910333
pythonidxpt:30866567
pythonnoceb:30805159
asynctok:30898717
pythontestfixt:30902429
pythonregdiag2:30936856
pythonmypyd1:30879173
pythoncet0:30885854
h48ei257:31000450
pythontbext0:30879054
accentitlementsc:30995553
dsvsc016:30899300
dsvsc017:30899301
dsvsc018:30899302
cppperfnew:31000557
dsvsc020:30976470
pythonait:31006305
chatpanelc:31048052
dsvsc021:30996838
f3je6385:31013174
pythoncenvpt:31049070
a69g1124:31046351
pythonprt:31047983
dwnewjupytercf:31046870
fa862199:31048878

```

</details>

**Comment by tomaszkapron** [comment]
Having similar issues

**Comment by javierjcf** [comment]
Same here

**Comment by edithli** [comment]
similar issues here on Mac

```sh
/private/var/folders/rd/pts24m5j6x35m7ytf_q60yn80000gn/T/AppTranslocation/A5742310-7FCA-4A76-9CE4-CA349B49206B/d/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --type=utility --utility-sub-type=node.mojom.NodeService --lang=en-US --service-sandbox-type=none --dns-result-order=ipv4first --inspect-port=0 --user-data-dir=/Users/lss/Library/Application Support/Code --standard-schemes=vscode-webview,vscode-file --enable-sandbox --secure-schemes=vscode-webview,vscode-file --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --code-cache-schemes=vscode-webview,vscode-file --shared-files --field-trial-handle=1718379636,r,11998688379030772291,12062321241314744751,262144 --enable-features=kWebSQLAccess --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --variations-seed-version
```

vs code version:
```
Version: 1.88.0 (Universal)
Commit: 5c3e652f63e798a5ac2f31ffd0d863669328dc4c
Date: 2024-04-03T13:28:18.899Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Darwin arm64 23.4.0
```

system info
```
macOS Sonoma Version 14.4.1 (23E224)
```

![image](https://github.com/microsoft/vscode/assets/13740057/de29656d-5b7b-4481-a665-402b2a671f20)

![image](https://github.com/microsoft/vscode/assets/13740057/d1e184d9-ee48-4d6f-8c65-b437ae28cd50)

**Comment by deepak1556** [comment]
That would be the extension host process, disable extensions will only have effect on installed extensions, builtin ones will be run and seems like an issue from one of them. Please use the following steps to capture a cpu profile of the process,

```
* Launch vscode like `code --inspect-extensions=9229`
* In chrome browser visit chrome://inspect/ or in edge browser visit edge://inspect
* You should see an entry under Remote Target
* If you don't see the above entry, then
  * Under Devices, click open dedicated devtools for node
  * Under Connection tab, add localhost:9229
  * You should now see the target
  * If there are multiple targets pick the one that matches the PID of the process showing high cpu
* Click inspect
* Go to Profiler tab and start the CPU profiler
* Stop the profiler after a few seconds and attach it to this issue.
```

**Comment by xrstf** [comment]
Same here.

Version: 1.89.1
Commit: dc96b837cf6bb4af9cd736aa3af08cf8279f7685
Date: 2024-05-07T05:16:23.416Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Linux x64 6.1.0-21-amd64

**Comment by petoknm** [comment]
The issue stopped happening on my machine, at least for now, so I can't perform the extra profiling at the moment.
However, I would really appreciate if people who do experience this issue, performed the profiling according to comment https://github.com/microsoft/vscode/issues/213136#issuecomment-2141179362 instead of just commenting that it happens. We need to give the team more data to debug this issue. Thank you.

**Comment by xrstf** [comment]
Updating to 1.90.1 did not solve the issue. The moment I have a workspace open, the file watcher process continues to use 100% of one CPU core.

**Comment by Geczy** [comment]
have the same issue, did not see anything in vscode process explorer showing an extension that was using high cpu. it just said extension host was high

> mgates           94478 138.3  0.1 1594142160  52704   ??  R     5Jun24 8461:12.92 /Applications/Visual Studio Code.app/Contents/Frameworks/Code Helper (Plugin).app/Contents/MacOS/Code Helper (Plugin) --type=utility --utility-sub-type=node.mojom.NodeService --lang=en-US --service-sandbox-type=none --dns-result-order=ipv4first --inspect-port=0 --user-data-dir=/Users/mgates/Library/Application Support/Code --standard-schemes=vscode-webview,vscode-file --enable-sandbox --secure-schemes=vscode-webview,vscode-file --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --code-cache-schemes=vscode-webview,vscode-file --shared-files --field-trial-handle=1718379636,r,7905944025835438358,16796389049385230011,262144 --enable-features=kWebSQLAccess --disable-features=CalculateNativeWinOcclusion,SpareRendererForSitePerProcess --variations-seed-version

**Comment by petoknm** [comment]
Please profile according to https://github.com/microsoft/vscode/issues/213136#issuecomment-2141179362

**Comment by Geczy** [comment]
@petoknm there is no profiler, but there is a "performance" tab that i can record with. let me know if i did it right. i can try again

200%+ cpu usage

[CPU-20240618T061059.cpuprofile](https://github.com/user-attachments/files/15885221/CPU-20240618T061059.cpuprofile)

**Comment by fuunnx** [comment]
Same story here, but different extension. [vscode-gitblame](https://github.com/waderyan/vscode-gitblame) is the culprit.

There is a lot of calls to async_hooks that causes a lot of computation (with `registerDestroyHook`). I guess those extensions were developed in a time before async_hooks were introduced. But now, async_hooks add a lot of overhead.

Here is the CPU profile for the curious :

[mojom-CPU-20240620T100137.cpuprofile](https://github.com/user-attachments/files/15910816/mojom-CPU-20240620T100137.cpuprofile)

**Comment by 3ng7n33r** [comment]
Same problem here since my last VSCode update to version 1.90.2. The culprit seems to be the dev-container extension for me. In addition to the CPU load, it also blocks a port that my application needs to function.
Killing the process unblocks the port and makes everything run smooth again until I restart the dev-container. After I kill it, VSCode presents me with an error message:
`A shared background process terminated unexpectedly. Please restart the application to recover.`
which can be ignored for now as vscode and the container seem to work completely fine.

**Comment by ravm** [comment]
Same issue here, except the CPU usage will stay at 100% unless I disable **all** of VSCode's built-in extensions.

**Comment by javierjcf** [comment]
I uninstall gitlens extension. I have long projects with a lot of git repositories inside. gitlens eatc my cpu and ram

**Comment by l-miskovsky** [comment]
Same issue with monorepo lerna project here. Seems that it's triggered by some monorepo + watcher process combination but definitely looks like a VSCode bug. I have to kill the offending process after running a command that includes file watching (`next dev`), CPU goes back to normal and the process that started this issue keeps running normally.

**Comment by petoknm** [comment]
Ok, I was finally able to repro this issue again, and I captured the CPU profile: 
[CPU-20240808T144337.cpuprofile](https://github.com/user-attachments/files/16548201/CPU-20240808T144337.cpuprofile)

In my case, I see it being mostly idle, but possibly doing some async stuff, not sure...

I see repeated calls to `getConfiguration` and `createSocket` but I have no idea where they are coming from.

VS Code version: Code 1.92.0 (b1c0a14de1414fcdaa400695b4db1c0799bc3124, 2024-07-31T23:26:45.634Z)
OS version: Linux x64 5.15.0-117-generic snap

**Comment by Geczy** [comment]
@petoknm try disabling extension `ms-vscode-remote.remote-containers` and run a profile again

**Comment by petoknm** [comment]
I also see a lot of git errors in the extension host output:
```
2024-08-08 14:38:28.124 [error] Error: Git error
	at ChildProcess.<anonymous> (/snap/code/165/usr/share/code/resources/app/extensions/git/dist/main.js:2:1030937)
	at ChildProcess.emit (node:events:531:35)
	at ChildProcess._handle.onexit (node:internal/child_process:294:12)
2024-08-08 14:38:28.124 [error] Error: Git error
	at ChildProcess.<anonymous> (/snap/code/165/usr/share/code/resources/app/extensions/git/dist/main.js:2:1030937)
	at ChildProcess.emit (node:events:531:35)
	at ChildProcess._handle.onexit (node:internal/child_process:294:12)
```
but it doesn't say what the error is, so I have no idea what is going on.
But what made me notice the CPU issue was that all git operations through vscode were unusably slow, and I actually still can't commit through vscode... its just showing the little loading line that moves horizontally...

**Comment by petoknm** [comment]
Another profile with remote-containers disabled: 
[CPU-20240808T152153.cpuprofile](https://github.com/user-attachments/files/16548622/CPU-20240808T152153.cpuprofile)

Looks totally idle to me from the profile, but top is showing CPU at 100 for process:
```
    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                   
1997401 pmajchr+  20   0 1131.9g   1.0g  70960 S 100.0   0.8   2:48.19 code  
```

and thats the mojom.NodeService thing when I look at the commandline of that process.

**Comment by petoknm** [comment]
```
$ time git status
...
real    0m7.303s
user    0m1.669s
sys     0m46.423s
```

yea, its a large project, but via terminal I can still work with git fine... but throught vscode it never shows me the "source control" stuff:
![image](https://github.com/user-attachments/assets/c5e7dfbc-cd70-4abe-b5cc-889361a11f91)

**Comment by macksal** [comment]
I found the same problem when an XCode build was much slower than usual on a new Macbook.

There were 4-5 VSCode processes each consuming 100-150% CPU: _Code Helper_, _Code Helper (Plugin)_ etc.

Modern OSX seems to keep the UI more responsive when processes are hogging CPU, so I didn't notice until checking Activity Monitor.

I followed the steps here, disabled many extensions, no change. I then quit VSCode and discovered that the processes did not die, so they were actually zombified. Force quit, restart Code and performance is back to normal.

**Comment by aatkin** [comment]
> I also see a lot of git errors in the extension host output:
> 
> ```
> 2024-08-08 14:38:28.124 [error] Error: Git error
> 	at ChildProcess.<anonymous> (/snap/code/165/usr/share/code/resources/app/extensions/git/dist/main.js:2:1030937)
> 	at ChildProcess.emit (node:events:531:35)
> 	at ChildProcess._handle.onexit (node:internal/child_process:294:12)
> 2024-08-08 14:38:28.124 [error] Error: Git error
> 	at ChildProcess.<anonymous> (/snap/code/165/usr/share/code/resources/app/extensions/git/dist/main.js:2:1030937)
> 	at ChildProcess.emit (node:events:531:35)
> 	at ChildProcess._handle.onexit (node:internal/child_process:294:12)
> ```
> 
> but it doesn't say what the error is, so I have no idea what is going on. But what made me notice the CPU issue was that all git operations through vscode were unusably slow, and I actually still can't commit through vscode... its just showing the little loading line that moves horizontally...

I am also experiencing unusually high CPU usage on MBP 2019 (with Intel chip) related to VSCode git extension. I didn't pick what errors were happening, but profiling showed that the git extension was calling something very often (almost every 20ms or so). Once I turned git extension off in VSCode settings, all the extra CPU usage went down. When I tried turning it back on, CPU usage went back up immediately. I also restarted system multiple times before and after finding out about the profiling results. I'm quite convinced that my issue is related to the git extension, but I don't know what could be the root cause.

**Comment by jasonbronson** [comment]
This is happening when the laptop goes into sleep mode and back out of it. The CPU spikes and stays at 100%.  Very annoying issue. Burns through a battery.

**Comment by xrstf** [comment]
Disabling GitLens _and restarting vscode_ has solved it for me. Re-enabling GitLens instantly makes one process use 100% CPU again.

**Comment by pierre-fromager** [comment]
It seems that bad files or directories permissions in your project is the issue

**Comment by HiFiPhile** [comment]
> Same story here, but different extension. [vscode-gitblame](https://github.com/waderyan/vscode-gitblame) is the culprit.

@fuunnx Hi, I've the same issue, in my case it happens on branch switching of a large python project. Have you found a better git blame extension ?

**Comment by fuunnx** [comment]
> > Same story here, but different extension. [vscode-gitblame](https://github.com/waderyan/vscode-gitblame) is the culprit.
> 
> @fuunnx Hi, I've the same issue, in my case it happens on branch switching of a large python project. Have you found a better git blame extension ?

I'm using GitLens now

**Comment by casual-lemon** [comment]
has there been any recent updates?

![Image](https://github.com/user-attachments/assets/8bbdbec3-473d-4cd1-a56f-9c81a5f0803c)

**Comment by casual-lemon** [comment]
![Image](https://github.com/user-attachments/assets/cb5ab7b6-d82a-469c-8339-22a75322c917)

**Comment by eliezerben** [comment]
I am facing this same issue with the october update

**Comment by GZTimeWalker** [comment]
Same here, on macOS 15.1, Extension Bisect will not solve this issue.

![Image](https://github.com/user-attachments/assets/01393df5-4a51-4a6b-bfc6-6b774db7c90d)

These process will not be killed after restart the extension host, this action will create more 100% process, and I can only use `killall -9 "Code Helper (Plugin)"` to solve this temporary.

**Comment by GZTimeWalker** [comment]
> I just started having this issue. Running v1.95.1

continue our issue in https://github.com/microsoft/vscode/issues/232699, this issue may not be relevant.

**Comment by ccforeverd** [comment]
same issue

```
Version: 0.42.4
VSCode Version: 1.93.1
Commit: b1e87884330fc271d5eb589e368c35f14e76dec0
Date: 2024-10-29T17:17:29.347Z
Electron: 30.4.0
Chromium: 124.0.6367.243
Node.js: 20.15.1
V8: 12.4.254.20-electron.0
OS: Darwin arm64 24.0.0
```

**Comment by Marinell0** [comment]
Having the same problem:

``` 
Version: 1.95.1
Commit: 65edc4939843c90c34d61f4ce11704f09d3e5cb6
Date: 2024-10-31T05:14:54.222Z
Electron: 32.2.1
ElectronBuildId: 10427718
Chromium: 128.0.6613.186
Node.js: 20.18.0
V8: 12.8.374.38-electron.0
OS: Linux x64 6.8.0-48-generic snap
```

This is extremely bad since in my workflow I open multiple windows at the same time. Each time I open a window a new process opens and uses 100% of a CPU core, bringing the PC to a halt.

I have to manually keep killing the process until I'm able to use the computer correctly.

**Comment by lyatziv** [comment]
Disabling GitLens did **not** work for me.

![Image](https://github.com/user-attachments/assets/362169fe-df02-4d32-b904-484de6df86a4)

Results form `code --status`
```
Version:          Code 1.95.2 (e8653663e8840adaf45af01eab5c627a5af81807, 2024-11-07T11:07:22.054Z)
OS Version:       Linux x64 6.9.3-76060903-generic
CPUs:             <-- Redacted -->
Memory (System):  62.66GB (38.90GB free)
Load (avg):       2, 2, 2
VM:               0%
Screen Reader:    no
Process Argv:     --crash-reporter-id c9512e1b-3771-4b99-8529-05fd315b2bf0
GPU Status:       2d_canvas:                              enabled
                  canvas_oop_rasterization:               enabled_on
                  direct_rendering_display_compositor:    disabled_off_ok
                  gpu_compositing:                        enabled
                  multiple_raster_threads:                enabled_on
                  opengl:                                 enabled_on
                  rasterization:                          enabled
                  raw_draw:                               disabled_off_ok
                  skia_graphite:                          disabled_off
                  video_decode:                           enabled
                  video_encode:                           disabled_software
                  vulkan:                                 disabled_off
                  webgl:                                  enabled
                  webgl2:                                 enabled
                  webgpu:                                 disabled_off
                  webnn:                                  disabled_off

CPU %   Mem MB     PID  Process
    0      128  2183622 code main
    0        0  2183627    zygote
    0      128  2183667      gpu-process
    0        0  2183628    zygote
    0        0  2183630      zygote
    0      385  2183685 window [1] (tt_carrier_receive_gen_mr.js - dev - Visual Studio Code)
    0       64  2183670    utility-network-service
    0       64  2184104 ptyHost
    0        0  2184525      /usr/bin/zsh -i
    0        0  2184529      /usr/bin/zsh -i
    0        0  2188744        sh /usr/bin/code --status
    0       64  2188749          electron-nodejs (cli.js )
    0      128  2188764            /usr/share/code/code --status
    0        0  2188767              zygote
    0       64  2188808                gpu-process
    0        0  2188768              zygote
    0        0  2188770                zygote
    0        0  2188811              utility-network-service
    0        0  2188825              /bin/sh -c /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0        0  2188826                /usr/bin/ps -ax -o pid=,ppid=,pcpu=,pmem=,command=
    0        0  2184531      /usr/bin/zsh -i
    0        0  2184536      /usr/bin/zsh -i
    0        0  2184538      /usr/bin/zsh -i
    0        0  2184540      /usr/bin/zsh -i
    0        0  2184545      /usr/bin/zsh -i
    0       64  2184151 shared-process
    0       64  2184152 fileWatcher [1]

Workspace Stats: 
|  Window (<-- Redacted --> - Visual Studio Code)
|    Folder (dev): 1215 files
|      File types: ts(385) js(233) json(153) sql(65) md(48) xml(33) sh(17)
|                  ftl(15) mmd(8) csv(7)
|      Conf files: tsconfig.json(20) package.json(18) settings.json(1)
|                  tasks.json(1) jsconfig.json(1) project.json(1)
```

**Comment by lyatziv** [comment]
Followup: Ran the Bisect extension function - Did **not** resolve the issue. :(

**Comment by Vystii** [comment]
same issue: code 1.94.2
i'm currently working with django and everything is ok as long as i don't open any template. 
I also tried to remove the gitlens extension(without really getting how it can be related)

**Comment by Vystii** [comment]
I disabled all my extensions related to django and enabled them afterward

**Comment by lyatziv** [comment]
So, I noticed that the TypeScript service was cranking out verbose logs. These included entire file contents. Setting the logging off verbose seems to alleviate the problem I was having.
--
Lavi 



Nov 29, 2024, 2:28 p.m. by ***@***.***:

**Comment by samuela** [comment]
I just ran into an issue matching this description and ended up tracking it down to the built-in json-language-features extension: https://github.com/microsoft/vscode/issues/237245

**Comment by dom-zhu** [comment]
I'm experiencing this, and tried to do an extension bisect, with it failing with all extensions off but still having the issue.

**Comment by lyatziv** [comment]
Thanks @JonKohler 

```
// settings.json
  "git.autofetch": false,
  "git.repositoryScanIgnoredFolders": [
    "node_modules",
    "./dist"
  ],
// ...

Before:
![Image](https://github.com/user-attachments/assets/51461fff-975d-4ff8-9c05-b170ae6029fa)

After:
![Image](https://github.com/user-attachments/assets/68f6600f-04c6-4f26-89c9-49db4d6d0e94)

**Comment by SolidTux** [comment]
That sadly did not fix the issue. I'm also using it in a repo that does not have node, but another folder (that's ignored for git) with a huge amount of files. I added this folder to the ignored folders and disable autofetch, but the high CPU usage remains.

**Comment by Kovinda** [comment]
> GitLens

GitLens was the culprit in my case too

**Comment by keelangibb** [comment]
For me I had 2 problematic extensions.
1. TxtSynax: which gave a problem only when opening really big files, no big deal
2. Error Lens: Problem can be reproduced by running a linter (using the command pallet) on a large code base. It will cause high cpu usage for minutes even if you had no open files when the command was run. Running the linter from the terminal won't trigger it. Extension bisect found both of these. I had no issues with GitLens.

**Comment by kiprasmel** [comment]
apart from the others mentioned already, i also had another surprising extension: Shan.code-settings-sync
https://github.com/shanalikhan/code-settings-sync

very weird since it's not actively used at all. and yet, each extension-host would be at ~30-40% utilization, after uninstalling, went down to ~1%

**Comment by gnpwaritschlager** [comment]
I've copied @Timmm's comment into a new feature #259225 to make the process explorer or the subprocess command line args more readable.

**Comment by iamqiz** [comment]
same issue, very high cpu usage!  node.mojom.NodeService

the process command line is : 
```
"C:\Users\pcmsi\AppData\Local\Programs\Microsoft VS Code\Code.exe" --type=utility --utility-sub-type=node.mojom.NodeService --lang=en-US --service-sandbox-type=none --video-capture-use-gpu-memory-buffer --user-data-dir="C:\Users\pcmsi\AppData\Roaming\Code" --standard-schemes=vscode-webview,vscode-file --enable-sandbox --secure-schemes=vscode-webview,vscode-file --cors-schemes=vscode-webview,vscode-file --fetch-schemes=vscode-webview,vscode-file --service-worker-schemes=vscode-webview --code-cache-schemes=vscode-webview,vscode-file --field-trial-handle=1572,i,17947856912118353302,3269798868404199376,262144 --enable-features=DocumentPolicyIncludeJSCallStacksInCrashReports,EarlyEstablishGpuChannel,EnableTransparentHwndEnlargement,EstablishGpuChannelAsync --disable-features=CalculateNativeWinOcclusion,FontationsLinuxSystemFonts,ScreenAIOCREnabled,SpareRendererForSitePerProcess,WinDelaySpellcheckServiceInit --variations-seed-version --mojo-platform-channel-handle=4164 /prefetch:8
```

**Comment by gfdb** [comment]
TLDR: make sure you add big directories (datasets, node_modules, etc) to your .gitignore, if you don't, vscode's source control will automatically try to track/index/process it all and spike your cpu usage.

just thought I'd comment on this thread since I managed to figure out what was causing it. This thread was the top google result when typing the first thing I saw in htop.

What I did: Copied a dataset directory into my repo for testing with 100GB+ of small files. It was not in .gitignore. 

What happened: The built in source control was trying to track/process all of these.

How I fixed it: Added it to .gitignore and cpu usage went back to normal. I would expect any directory with a large amount of data or number of files to also cause this issue e.g. node_modules directory for js devs

I use the terminal for git, so I didn't even know about the built in source control, which made debugging this quite annoying. In the end, source control was doing what it was supposed to do since this dataset directory wasn't in my .gitignore, however it might be worth considering adding a check for situations like this. For example, when you open large files in vscode, there is the "The file is not displayed in the text editor because it is very large" message. Maybe a similar warning/message for source control tracking enormous directories would be good...

**Comment by manuel-wagesreither** [comment]
I'm affected as well although I did just about everything to help it.

* Entirely fresh install of VS Code with just the default set of extensions. (It says `0` extensions installed.)
* Added every single directory to the `.gitignore`.
* Made the project a non-trusted project, which disables a few built-in extensions.
* In the workspace-trust dialogue, there's a link you can click, that tells you which extensions would not work in a non-trusted project. Clicking that link lists a few extensions that were not shown before. I disabled each of them.

**There's literally no way I can use Visual Studio Code without saturating one cpu core.**

I tried Codium, but it's the same there as well. I have no other option than to change to another editor.

How is this not affecting everyone? There's nothing special about my system.

```
Version: 1.108.2
Commit: c9d77990917f3102ada88be140d28b038d1dd7c7
Date: 2026-01-21T13:52:09.270Z
Electron: 39.2.7
ElectronBuildId: 12953945
Chromium: 142.0.7444.235
Node.js: 22.21.1
V8: 14.2.231.21-electron.0
OS: Linux x64 6.12.63+deb13-amd64
```
I'm on `Debian GNU/Linux 13 (trixie)`.
